internal/mq/msgstream/mqwrapper/kafka/kafka_consumer.go

```diff
@@ -0,0 +1,140 @@
+package kafka
+
+import (
+	"sync"
+	"time"
+
+	"github.com/confluentinc/confluent-kafka-go/kafka"
+	"github.com/milvus-io/milvus/internal/log"
+	"github.com/milvus-io/milvus/internal/mq/msgstream/mqwrapper"
+	"go.uber.org/zap"
+)
+
+type Consumer struct {
+	c          *kafka.Consumer
+	msgChannel chan mqwrapper.Message
+	closeFlag  bool
+	skipMsg    bool
+	topicName  string
+	groupID    string
+	closeCh    chan struct{}
+	chanOnce   sync.Once
+	closeOnce  sync.Once
+	wg         sync.WaitGroup
+}
+
+func newKafkaConsumer(consumer *kafka.Consumer, topicName string, groupID string) *Consumer {
+	msgChannel := make(chan mqwrapper.Message, 1)
+	closeCh := make(chan struct{})
+	skipMsg := false
+
+	kafkaConsumer := &Consumer{c: consumer,
+		msgChannel: msgChannel,
+		skipMsg:    skipMsg,
+		topicName:  topicName,
+		groupID:    groupID,
+		closeCh:    closeCh,
+	}
+
+	return kafkaConsumer
+}
+
+func (kc *Consumer) Subscription() string {
+	return kc.groupID
+}
+
+// Chan provides a channel to read consumed message.
+// There are some illustrations need to clarify,
+// 1.confluent-kafka-go recommend us to use function-based consumer,
+// channel-based consumer API had already deprecated, see more details
+// https://github.com/confluentinc/confluent-kafka-go.
+//
+// 2.This API of other MQ return channel directly, but it depends on
+// readMessage firstly which means it be always triggered within select-case
+// invocation. However, it still works well, because it covers all messages
+// consume situation: start from the earliest or latest position to keep consume;
+// start from a seek position to specified end position.
+func (kc *Consumer) Chan() <-chan mqwrapper.Message {
+	if kc.skipMsg {
+		msg := kc.readMessage()
+		if msg != nil {
+			kc.skipMsg = false
+		}
+	}
+
+	msg := kc.readMessage()
+
+	if msg != nil {
+		kc.msgChannel <- &kafkaMessage{msg: msg}
+	} else {
+		kc.msgChannel <- nil
+	}
+
+	return kc.msgChannel
+}
+
+func (kc *Consumer) readMessage() *kafka.Message {
+	msg, err := kc.c.ReadMessage(1 * time.Second)
```

Comment:
why do we not use -1 for indefinite wait?

Reply:
As I mentioned, the block is another strategy, but I use a limitation timeout, return nil value if the timeout. there is an advantage to provide a more flexible API operation by preventing infinite wait